open_flamingo/train/train_utils.py

```diff
@@ -3,282 +3,205 @@
-        labels[labels == tokenizer.pad_token_id] = -100
-        labels[labels == tokenizer.eos_token] = -100
-        labels[labels == media_token_id] = -100
-        labels = labels.to(device_id)
-
-        # gradient accumulation w/ fsdp cpu offloading requires a no_sync context manager
-        with autocast():
-            loss_laion = model(
-                vision_x=images,
-                lang_x=input_ids,
+            # forward pass
+            dataset_loss = compute_loss_fn(
+                model=model,
+                tokenizer=tokenizer,
+                images=images,
+                input_ids=input_ids,
                 attention_mask=attention_mask,
-                labels=labels,
-            )[0]
-
-        divided_loss_laion = loss_laion / args.gradient_accumulation_steps
-        (divided_loss_laion * args.loss_multiplier_laion).backward()
-
-        #### MMC4 FORWARD PASS ####
-        images = batch_mmc4[0].to(device_id, dtype=cast_dtype, non_blocking=True)
-        images = rearrange(images, "b (t f) c h w -> b t f c h w", f=1)
-        input_ids = torch.stack([x[0] for x in batch_mmc4[1]]).squeeze(1)
-        attention_mask = torch.stack([x[1] for x in batch_mmc4[1]]).squeeze(1)
-
-        # set up labels; language model is expected to handle shifting
-        labels = input_ids.clone()
-        labels[labels == tokenizer.pad_token_id] = -100
-        labels[labels == tokenizer.eos_token] = -100
-        for i in range(labels.shape[0]):
-            # remove loss for any token before the first <image> token
-            label_idx = 0
-            while (
-                label_idx < labels.shape[1] and labels[i][label_idx] != media_token_id
-            ):
-                labels[i][label_idx] = -100
-                label_idx += 1
-
-            # get index of all endofchunk tokens in the sequence
-            endofchunk_idxs = torch.where(labels[i] == endofchunk_token_id)[0]
-            for endofchunk_idx in endofchunk_idxs:
-                token_idx = endofchunk_idx + 1
-                while (
-                    token_idx < labels.shape[1]
-                    and labels[i][token_idx] != media_token_id
-                ):
-                    labels[i][token_idx] = -100
-                    token_idx += 1
-
-        labels[labels == media_token_id] = -100
-        labels = labels.to(device_id)
-
-        # gradient accumulation w/ fsdp cpu offloading requires a no_sync context manager
-        with autocast():
-            loss_mmc4 = model(
-                vision_x=images,
-                lang_x=input_ids.to(device_id),
-                attention_mask=attention_mask.to(device_id),
-                labels=labels,
-            )[0]
+                autocast=autocast,
+            )
 
             # if loss is nan, skip this batch
             # this hack of skipping the batch is not FSDP-compatible
-            if torch.isnan(loss_mmc4):
+            if torch.isnan(dataset_loss):
```

Comment:
We previously fixed the NAN issue on the data side right? Since this call will crash fsdp/deepspeed runs so should we remove?

Reply:
Yeah, sounds good! I'm not sure what we'll see at larger scales, but we can look for fsdp-compatible solutions then.